node/libs/crypto/src/bn254/testonly.rs

```diff
@@ -1,33 +1,50 @@
 //! Random key generation, intended for use in testing
 
-use super::{AggregateSignature, PublicKey, SecretKey, Signature};
+use pairing::bn256::{Fr, G1, G2};
 use rand::{distributions::Standard, prelude::Distribution, Rng};
+use rand04::SeedableRng;
+use rand04::{Rand, StdRng as rand04StdRng};
+
+use super::{AggregateSignature, PublicKey, SecretKey, Signature};
 
 /// Generates a random SecretKey. This is meant for testing purposes.
 impl Distribution<SecretKey> for Standard {
     fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> SecretKey {
-        let rand = ark_bn254::Fr::new(rng.gen());
-        SecretKey(rand)
+        let mut rng = rng08_to_rng04(rng);
+        let scalar = Fr::rand(&mut rng);
+        SecretKey(scalar)
     }
 }
 
 /// Generates a random PublicKey. This is meant for testing purposes.
 impl Distribution<PublicKey> for Standard {
     fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> PublicKey {
-        PublicKey(rng.gen())
+        let mut rng = rng08_to_rng04(rng);
+        let p = G2::rand(&mut rng);
+        PublicKey(p)
     }
 }
 
 /// Generates a random Signature. This is meant for testing purposes.
 impl Distribution<Signature> for Standard {
     fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Signature {
-        Signature(rng.gen())
+        let mut rng = rng08_to_rng04(rng);
+        let p = G1::rand(&mut rng);
+        Signature(p)
     }
 }
 
 /// Generates a random AggregateSignature. This is meant for testing purposes.
 impl Distribution<AggregateSignature> for Standard {
     fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> AggregateSignature {
-        AggregateSignature(rng.gen())
+        let mut rng = rng08_to_rng04(rng);
+        let p = G1::rand(&mut rng);
+        AggregateSignature(p)
     }
 }
+
+fn rng08_to_rng04<R: Rng + ?Sized>(rng: &mut R) -> rand04StdRng {
+    let val = rng.gen::<usize>();
+    let seed: &[usize] = std::slice::from_ref(&val);
+    rand04::StdRng::from_seed(seed)
```

Comment:
I'd expect to see a wrapper for `rand08::RngCore` that implements `rand04::Rng`; this should be trivial, since the latter requires [a single method](https://docs.rs/rand/0.4.6/rand/trait.Rng.html#tymethod.next_u32), which is [readily available](https://docs.rs/rand/latest/rand/trait.RngCore.html#tymethod.next_u32) in `rand08::RngCore`.

So something like

```rust
struct RngWrapper<R>(R);

impl<R: RngCore> rand04::Rng for RngWrapper<R> {
    // ...
}

// Then, you can use it like
let p = G1::rand(&mut RngWrapper(rng));
```

Reply:
Good idea, thanks!